File: amazon/207_course_schedule.go

//dfs
func canFinish(numCourses int, prerequisites [][]int) bool{
	//building adjency list
	adj := make([][]int, numCourses)
	for _, pair:= range prerequisites{
		adj[pair[1]] = append(adj[pair[1]], pair[0])
	}

	//create a visited array
	visited := make([]int, numCourses)
	//defin DFS function for cycle detection
	var hasCycle func(int) bool
	hasCycle = func(v int) bool{
		if visited[v] ==1 {
			return true
		}
		if visited[v]==2{
			return false
		}
		visited[v] =1

		for _, neighbor:= range adj[v]{
			if hasCycle(neighbor){
				return true
			}
		}
		visited[v] = 2
		return false
	}
	//chech each coruse for cycles
	for i := 0; i < numCourses; i++{
		if visited[i] == 0{
			if hasCycle(i){
				return false
			}
		}
	}
	return true
}


//bfs via Kahn's ALgorithm

import "container/list"

func canFinish(numCourses int, prerequisites [][]int) bool{
	//build adjacency list and in-degree array
	adj := make([][]int, numCourses)
	inDegree := make([]int, numCourses)

	for _, pair:= range prerequisites{
		adj[pair[1]] = append(adj[pair[1]], pair[0])
		inDegree[pair[0]]++
	}
	//initialise the queue with courses that have in-degree of 0
	queue:= list.New()
	for i:= 0 ; i < numCourses; i++{
		if inDegree[i] == 0{
		queue.PushBack(i)
		}
	}

	processedCourses := 0
	for queue.Len() > 0{
		course := queue.Remove(queue.Front()).(int)
		processedCourses++

		for _, neighbor := range adj[course]{
			inDegree[neighbor]--
			if inDegree[neighbor] == 0 {
				queue.PushBack(neighbor)
			}
		}
	}
	return processedCourses == numCourses
}

File: amazon/207_course_schedule.py


#DFS- based solution(cycle detection) 
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        #create adjacency list for the graph 
        adj = [[] for _ in range(numCourses)]
        for dest, src in prerequisites:
            adj[src].append(dest)

        #stages 0 = unvisited, 1 = visiting, 2= visited
        visited = [0] * numCourses

        def hasCycle(v):
            if visited[v] == 1: #found a cycle       
                return True
            if visited[v] == 2:
                return False

            visited[v] = 1

            for neighbor in adj[v]:
                if hasCycle(neighbor):
                    return True

            visited[v] = 1

            for neighbor in adj[v]:
                if hasCycle(neighbor):
                    return True

            visited[v] =2
            return False

        for i in range(numCourses):
            if visited[i] == 0:
                if hasCycle(i):
                    return False

        return True

#DFS (Kahn's Algorithm) efficiency++ for the case.
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]])->bool:
        #create an adjacency list and calculate in-degrees
        adj = [[] for _ in range(numCourses)]
        inDegree = [0] * numCourses

        for dest, src in prerequisites:
            adj[src].append(dest)
            inDegree[dest] += 1

        #initilize the queue with all courses that have in degree of 0
        queue = deque()
        for i in range(numCourses):
            if inDegree[i] == 0:
                queue.append(i)

        #process the queue
        coursesProcessed = 0
        while queue:
            course = queue.popleft()
            coursesProcessed +=1

            for neighbor in adj[course]:
                inDegree[neighbor] -=1
                if inDegree[neighbor] == 0:
                    queue.append(neighbor)

        #rcheck if we were able to process all courses
        return coursesProcessed == numCourses